src/validator/CallPermitValidator.sol

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import {Validator} from "src/validator/Validator.sol";
+import {UserOperation} from "src/lib/ERC4337/utils/UserOperation.sol";
+import {Operations} from "src/lib/Operations.sol";
+import {Access} from "src/access/Access.sol";
+import {SignatureChecker} from "openzeppelin-contracts/utils/cryptography/SignatureChecker.sol";
+import {ECDSA} from "openzeppelin-contracts/utils/cryptography/ECDSA.sol";
+
+/// @dev Validator module that restricts valid signatures to only come from addresses
+/// that have been granted the `CALL_PERMIT` permission in the calling Accounts contract 
+contract CallPermitValidator is Validator {
+
+    constructor(address _entryPointAddress) Validator(_entryPointAddress) {}
+
+    /// @dev This example contract would be forwarded regular ECDSA signatures
+    function validateUserOp(
+        UserOperation calldata userOp, 
+        bytes32 userOpHash, 
+        uint256 missingAccountFunds
+    ) external virtual returns (uint256 validationData) {
+        // silence compiler by discarding unused variable
+        (missingAccountFunds);
+
+        /// @notice BLS sig aggregator and timestamp expiry are not currently supported by this contract 
+        /// so `bytes20(0x0)` and `bytes6(0x0)` suffice. To enable support for aggregator and timestamp expiry,
+        /// override the following params
+        bytes20 authorizer;
+        bytes6 validUntil;
+        bytes6 validAfter;
+        uint256 successData = uint256(bytes32(abi.encodePacked(authorizer, validUntil, validAfter)));
+
+        // terminate if recovered signer address does not match `userOp.sender`
+        if (!SignatureChecker.isValidSignatureNow(userOp.sender, userOpHash, userOp.signature)) return SIG_VALIDATION_FAILED;
+        
+        // apply this validator's authentication logic
+        bool validSigner = _verifySigner(userOp.sender);

```

Comment:
Confused why we are doing auth against `userOp.sender`? The sender is the contract that the validator is getting called from to perform the modular authorization. The `userOp.signature` should be assumed to be of the form `abi.encode(signer, nestedSignature)` which we need to decode and then check this same logic on the parsed signer:
```suggestion
        (address signer, bytes nestedSig) = abi.decode(userOp.signature, (addres, bytes))
        
        // terminate if signer does not match signature
        if (!SignatureChecker.isValidSignatureNow(signer, userOpHash, nestedSig)) return SIG_VALIDATION_FAILED;
        
        // apply this validator's authentication logic
        bool validSigner = _verifySigner(signer);

```

Reply:
I think I must have misunderstood whether or not we wanted the `signer` to be a part of the signature schema. 

I took the comments below to mean that we did not want it encoded into the signature as it was before, but it looks like I misinterpreted.

![image](https://github.com/0xStation/0xrails/assets/80549215/5da170b7-d827-44a9-a94d-e975b45213c8)

![image](https://github.com/0xStation/0xrails/assets/80549215/1f870a2b-a764-4ba1-a0b5-b6d28611083c)

Let's sync again on this to be sure

Reply:
Resolved this in 03f6b7e

It became clear that we would need to be craft our signature schema using `abi.encodePacked()` or simple string concatenation when creating them offchain. 

Sorry to flip flop on the encoding style: `abi.encode()` became an issue when trying to remain agnostic to the calldata after the first `VALIDATOR_FLAG` word, due to the schema mixing dynamic and static types.